// bling-erp-api/lib/entities/ctes.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const all_1 = __importDefault(require("../core/functions/all"));
const find_1 = __importDefault(require("../core/functions/find"));
const findBy_1 = __importDefault(require("../core/functions/findBy"));
const create_1 = __importDefault(require("../core/functions/create"));
const update_1 = __importDefault(require("../core/functions/update"));
const delete_1 = __importDefault(require("../core/functions/delete"));
function Ctes(api, raw) {
    const config = {
        api,
        raw,
        singularName: 'cte',
        pluralName: 'ctes'
    };
    const find = (numero, serie, options) => __awaiter(this, void 0, void 0, function* () {
        const findMethod = new find_1.default(config);
        const raw = options && options.raw !== undefined ? options.raw : config.raw;
        // @TODO: see how to reuse the code below
        if (options) {
            if (raw) {
                return yield findMethod.find(`${numero}/${serie}`, {
                    params: options.params,
                    raw: true
                });
            }
            else {
                return yield findMethod.find(`${numero}/${serie}`, {
                    params: options.params,
                    raw: false
                });
            }
        }
        else {
            return yield findMethod.find(`${numero}/${serie}`);
        }
    });
    const create = (data, options) => __awaiter(this, void 0, void 0, function* () {
        const createMethod = new create_1.default(config);
        const raw = options && options.raw !== undefined ? options.raw : config.raw;
        // @TODO: see how to reuse the code below
        if (options) {
            if (raw) {
                return yield createMethod.create(data, { raw: true }, options.loja);
            }
            else {
                return yield createMethod.create(data, { raw: false }, options.loja);
            }
        }
        else {
            return yield createMethod.create(data, { raw: false });
        }
    });
    const post = (id, options) => __awaiter(this, void 0, void 0, function* () {
        const createMethod = new create_1.default(Object.assign(Object.assign({}, config), { endpoint: `/cte/lancamento/contas/${id}` }));
        const raw = options && options.raw !== undefined ? options.raw : config.raw;
        // @TODO: see how to reuse the code below
        if (raw) {
            return yield createMethod.create(undefined, { raw: true });
        }
        else {
            return yield createMethod.create(undefined, { raw: false });
        }
    });
    const customDelete = (id, options) => __awaiter(this, void 0, void 0, function* () {
        const deleteMethod = new delete_1.default(Object.assign(Object.assign({}, config), { endpoint: '/cte/estorno/contas' }));
        const raw = options && options.raw !== undefined ? options.raw : config.raw;
        // @TODO: see how to reuse the code below
        if (raw) {
            return yield deleteMethod.delete(id, { raw: true });
        }
        else {
            return yield deleteMethod.delete(id, { raw: false });
        }
    });
    return Object.assign(config, {
        all: new all_1.default().all,
        find,
        findBy: new findBy_1.default().findBy,
        create,
        post,
        update: new update_1.default().update,
        delete: customDelete
    });
}
exports.default = Ctes;
