// bling-erp-api/lib/entities/products.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const all_1 = __importDefault(require("../core/functions/all"));
const find_1 = __importDefault(require("../core/functions/find"));
const findBy_1 = __importDefault(require("../core/functions/findBy"));
const create_1 = __importDefault(require("../core/functions/create"));
const delete_1 = __importDefault(require("../core/functions/delete"));
function Products(api, raw) {
    const config = {
        api,
        raw,
        singularName: 'produto',
        pluralName: 'produtos'
    };
    const update = (id, data, options) => __awaiter(this, void 0, void 0, function* () {
        const createMethod = new create_1.default(Object.assign(Object.assign({}, config), { endpoint: `${config.singularName}/${id}` }));
        const raw = options && options.raw !== undefined ? options.raw : config.raw;
        if (raw) {
            return yield createMethod.create(data, { raw: true });
        }
        else {
            return yield createMethod.create(data, { raw: false });
        }
    });
    const findBySupplierCode = (code, supplierId, options) => __awaiter(this, void 0, void 0, function* () {
        const findMethod = new find_1.default(config);
        const raw = options && options.raw !== undefined ? options.raw : config.raw;
        // @TODO: see how to reuse the code below
        if (options) {
            if (raw) {
                return yield findMethod.find(`${code}/${supplierId}`, {
                    params: options.params,
                    raw: true
                });
            }
            else {
                return yield findMethod.find(`${code}/${supplierId}`, {
                    params: options.params,
                    raw: false
                });
            }
        }
        else {
            return yield findMethod.find(`${code}/${supplierId}`);
        }
    });
    return Object.assign(config, {
        all: new all_1.default().all,
        find: new find_1.default().find,
        findBy: new findBy_1.default()
            .findBy,
        create: new create_1.default().create,
        update,
        delete: new delete_1.default().delete,
        findBySupplierCode
    });
}
exports.default = Products;
