// bling-erp-api/lib/entities/productionOrders.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const all_1 = __importDefault(require("../core/functions/all"));
const find_1 = __importDefault(require("../core/functions/find"));
const findBy_1 = __importDefault(require("../core/functions/findBy"));
const create_1 = __importDefault(require("../core/functions/create"));
const delete_1 = __importDefault(require("../core/functions/delete"));
function ProductionOrders(api, raw) {
    const config = {
        api,
        raw,
        singularName: 'ordemproducao',
        pluralName: 'ordensproducao'
    };
    return Object.assign(config, {
        all: new all_1.default().all,
        find: new find_1.default().find,
        findBy: new findBy_1.default().findBy,
        create: new create_1.default()
            .create,
        delete: new delete_1.default().delete
    });
}
exports.default = ProductionOrders;
