// bling-erp-api/lib/entities/purchaseOrders.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const all_1 = __importDefault(require("../core/functions/all"));
const find_1 = __importDefault(require("../core/functions/find"));
const findBy_1 = __importDefault(require("../core/functions/findBy"));
const create_1 = __importDefault(require("../core/functions/create"));
const update_1 = __importDefault(require("../core/functions/update"));
function PurchaseOrders(api, raw) {
    const config = {
        api,
        raw,
        singularName: 'pedidocompra',
        pluralName: 'pedidoscompra'
    };
    return Object.assign(config, {
        all: new all_1.default().all,
        find: new find_1.default().find,
        findBy: new findBy_1.default().findBy,
        create: new create_1.default().create,
        update: new update_1.default().update
    });
}
exports.default = PurchaseOrders;
