// bling-erp-api/lib/bling.js
'use strict';
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _Bling_api, _Bling_raw, _Bling_borderos, _Bling_customizedFields, _Bling_categories, _Bling_commercialProposals, _Bling_contacts, _Bling_deposits, _Bling_orders, _Bling_products, _Bling_purchaseOrders, _Bling_invoices, _Bling_shopCategories, _Bling_billsToPay, _Bling_billsToReceive, _Bling_contracts, _Bling_ctes, _Bling_paymentMethods, _Bling_productGroups, _Bling_nfces, _Bling_serviceInvoices;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bling = void 0;
const borderos_1 = __importDefault(require("./entities/borderos"));
const customizedFields_1 = __importDefault(require("./entities/customizedFields"));
const categories_1 = __importDefault(require("./entities/categories"));
const commercialProposals_1 = __importDefault(require("./entities/commercialProposals"));
const contacts_1 = __importDefault(require("./entities/contacts"));
const deposits_1 = __importDefault(require("./entities/deposits"));
const products_1 = __importDefault(require("./entities/products"));
const orders_1 = __importDefault(require("./entities/orders"));
const purchaseOrders_1 = __importDefault(require("./entities/purchaseOrders"));
const invoices_1 = __importDefault(require("./entities/invoices"));
const shopCategories_1 = __importDefault(require("./entities/shopCategories"));
const billsToPay_1 = __importDefault(require("./entities/billsToPay"));
const billsToReceive_1 = __importDefault(require("./entities/billsToReceive"));
const contracts_1 = __importDefault(require("./entities/contracts"));
const ctes_1 = __importDefault(require("./entities/ctes"));
const paymentMethods_1 = __importDefault(require("./entities/paymentMethods"));
const productGroups_1 = __importDefault(require("./entities/productGroups"));
const nfces_1 = __importDefault(require("./entities/nfces"));
const serviceInvoices_1 = __importDefault(require("./entities/serviceInvoices"));
const createError_1 = __importDefault(require("./core/helpers/createError"));
const qs = __importStar(require("querystring"));
const axios_1 = __importDefault(require("axios"));
class Bling {
    constructor(apiKey, options = { raw: false }) {
        _Bling_api.set(this, void 0);
        _Bling_raw.set(this, void 0);
        _Bling_borderos.set(this, void 0);
        _Bling_customizedFields.set(this, void 0);
        _Bling_categories.set(this, void 0);
        _Bling_commercialProposals.set(this, void 0);
        _Bling_contacts.set(this, void 0);
        _Bling_deposits.set(this, void 0);
        _Bling_orders.set(this, void 0);
        _Bling_products.set(this, void 0);
        _Bling_purchaseOrders.set(this, void 0);
        _Bling_invoices.set(this, void 0);
        _Bling_shopCategories.set(this, void 0);
        _Bling_billsToPay.set(this, void 0);
        _Bling_billsToReceive.set(this, void 0);
        _Bling_contracts.set(this, void 0);
        _Bling_ctes.set(this, void 0);
        _Bling_paymentMethods.set(this, void 0);
        _Bling_productGroups.set(this, void 0);
        _Bling_nfces.set(this, void 0);
        _Bling_serviceInvoices.set(this, void 0);
        if (!apiKey || typeof apiKey !== 'string') {
            throw (0, createError_1.default)({
                name: 'BlingModuleError',
                message: "The API key wasn't correctly provided for Bling connection.",
                status: '500',
                data: {
                    apiKey: apiKey
                },
                code: 'ERR_NO_API_KEY'
            });
        }
        if (typeof options.raw !== 'boolean') {
            throw (0, createError_1.default)({
                name: 'BlingModuleError',
                message: 'The raw attribute must be a boolean to configure the Bling connection.',
                status: '500',
                data: {
                    raw: options.raw
                },
                code: 'ERR_WRONG_BLING_RAW_ATTR'
            });
        }
        __classPrivateFieldSet(this, _Bling_raw, options.raw, "f");
        const api = axios_1.default.create({
            baseURL: 'https://bling.com.br/Api/v2'
        });
        api.interceptors.request.use((config) => {
            if (config.method &&
                ['POST', 'PUT', 'post', 'put'].includes(config.method)) {
                config.data = qs.stringify(Object.assign({ apikey: apiKey }, config.data));
            }
            config.params = Object.assign({ apikey: apiKey }, config.params);
            return config;
        });
        __classPrivateFieldSet(this, _Bling_api, api, "f");
    }
    static create(apiKey, options = { raw: false }) {
        return new this(apiKey, options);
    }
    borderos() {
        if (!__classPrivateFieldGet(this, _Bling_borderos, "f")) {
            __classPrivateFieldSet(this, _Bling_borderos, (0, borderos_1.default)(__classPrivateFieldGet(this, _Bling_api, "f"), __classPrivateFieldGet(this, _Bling_raw, "f")), "f");
        }
        return __classPrivateFieldGet(this, _Bling_borderos, "f");
    }
    customizedFields() {
        if (!__classPrivateFieldGet(this, _Bling_customizedFields, "f")) {
            __classPrivateFieldSet(this, _Bling_customizedFields, (0, customizedFields_1.default)(__classPrivateFieldGet(this, _Bling_api, "f"), __classPrivateFieldGet(this, _Bling_raw, "f")), "f");
        }
        return __classPrivateFieldGet(this, _Bling_customizedFields, "f");
    }
    camposCustomizados() {
        return this.customizedFields();
    }
    categories() {
        if (!__classPrivateFieldGet(this, _Bling_categories, "f")) {
            __classPrivateFieldSet(this, _Bling_categories, (0, categories_1.default)(__classPrivateFieldGet(this, _Bling_api, "f"), __classPrivateFieldGet(this, _Bling_raw, "f")), "f");
        }
        return __classPrivateFieldGet(this, _Bling_categories, "f");
    }
    categorias() {
        return this.categories();
    }
    shopCategories() {
        if (!__classPrivateFieldGet(this, _Bling_shopCategories, "f")) {
            __classPrivateFieldSet(this, _Bling_shopCategories, (0, shopCategories_1.default)(__classPrivateFieldGet(this, _Bling_api, "f"), __classPrivateFieldGet(this, _Bling_raw, "f")), "f");
        }
        return __classPrivateFieldGet(this, _Bling_shopCategories, "f");
    }
    categoriasLoja() {
        return this.shopCategories();
    }
    contacts() {
        if (!__classPrivateFieldGet(this, _Bling_contacts, "f")) {
            __classPrivateFieldSet(this, _Bling_contacts, (0, contacts_1.default)(__classPrivateFieldGet(this, _Bling_api, "f"), __classPrivateFieldGet(this, _Bling_raw, "f")), "f");
        }
        return __classPrivateFieldGet(this, _Bling_contacts, "f");
    }
    contatos() {
        return this.contacts();
    }
    billsToPay() {
        if (!__classPrivateFieldGet(this, _Bling_billsToPay, "f")) {
            __classPrivateFieldSet(this, _Bling_billsToPay, (0, billsToPay_1.default)(__classPrivateFieldGet(this, _Bling_api, "f"), __classPrivateFieldGet(this, _Bling_raw, "f")), "f");
        }
        return __classPrivateFieldGet(this, _Bling_billsToPay, "f");
    }
    contasAPagar() {
        return this.billsToPay();
    }
    billsToReceive() {
        if (!__classPrivateFieldGet(this, _Bling_billsToReceive, "f")) {
            __classPrivateFieldSet(this, _Bling_billsToReceive, (0, billsToReceive_1.default)(__classPrivateFieldGet(this, _Bling_api, "f"), __classPrivateFieldGet(this, _Bling_raw, "f")), "f");
        }
        return __classPrivateFieldGet(this, _Bling_billsToReceive, "f");
    }
    contasAReceber() {
        return this.billsToReceive();
    }
    contracts() {
        if (!__classPrivateFieldGet(this, _Bling_contracts, "f")) {
            __classPrivateFieldSet(this, _Bling_contracts, (0, contracts_1.default)(__classPrivateFieldGet(this, _Bling_api, "f"), __classPrivateFieldGet(this, _Bling_raw, "f")), "f");
        }
        return __classPrivateFieldGet(this, _Bling_contracts, "f");
    }
    contratos() {
        return this.contracts();
    }
    ctes() {
        if (!__classPrivateFieldGet(this, _Bling_ctes, "f")) {
            __classPrivateFieldSet(this, _Bling_ctes, (0, ctes_1.default)(__classPrivateFieldGet(this, _Bling_api, "f"), __classPrivateFieldGet(this, _Bling_raw, "f")), "f");
        }
        return __classPrivateFieldGet(this, _Bling_ctes, "f");
    }
    deposits() {
        if (!__classPrivateFieldGet(this, _Bling_deposits, "f")) {
            __classPrivateFieldSet(this, _Bling_deposits, (0, deposits_1.default)(__classPrivateFieldGet(this, _Bling_api, "f"), __classPrivateFieldGet(this, _Bling_raw, "f")), "f");
        }
        return __classPrivateFieldGet(this, _Bling_deposits, "f");
    }
    depositos() {
        return this.deposits();
    }
    paymentMethods() {
        if (!__classPrivateFieldGet(this, _Bling_paymentMethods, "f")) {
            __classPrivateFieldSet(this, _Bling_paymentMethods, (0, paymentMethods_1.default)(__classPrivateFieldGet(this, _Bling_api, "f"), __classPrivateFieldGet(this, _Bling_raw, "f")), "f");
        }
        return __classPrivateFieldGet(this, _Bling_paymentMethods, "f");
    }
    formasDePagamento() {
        return this.paymentMethods();
    }
    productGroups() {
        if (!__classPrivateFieldGet(this, _Bling_productGroups, "f")) {
            __classPrivateFieldSet(this, _Bling_productGroups, (0, productGroups_1.default)(__classPrivateFieldGet(this, _Bling_api, "f"), __classPrivateFieldGet(this, _Bling_raw, "f")), "f");
        }
        return __classPrivateFieldGet(this, _Bling_productGroups, "f");
    }
    grupoDeProdutos() {
        return this.productGroups();
    }
    nfces() {
        if (!__classPrivateFieldGet(this, _Bling_nfces, "f")) {
            __classPrivateFieldSet(this, _Bling_nfces, (0, nfces_1.default)(__classPrivateFieldGet(this, _Bling_api, "f"), __classPrivateFieldGet(this, _Bling_raw, "f")), "f");
        }
        return __classPrivateFieldGet(this, _Bling_nfces, "f");
    }
    invoices() {
        if (!__classPrivateFieldGet(this, _Bling_invoices, "f")) {
            __classPrivateFieldSet(this, _Bling_invoices, (0, invoices_1.default)(__classPrivateFieldGet(this, _Bling_api, "f"), __classPrivateFieldGet(this, _Bling_raw, "f")), "f");
        }
        return __classPrivateFieldGet(this, _Bling_invoices, "f");
    }
    notasFiscais() {
        return this.invoices();
    }
    serviceInvoices() {
        if (!__classPrivateFieldGet(this, _Bling_serviceInvoices, "f")) {
            __classPrivateFieldSet(this, _Bling_serviceInvoices, (0, serviceInvoices_1.default)(__classPrivateFieldGet(this, _Bling_api, "f"), __classPrivateFieldGet(this, _Bling_raw, "f")), "f");
        }
        return __classPrivateFieldGet(this, _Bling_serviceInvoices, "f");
    }
    notasServicos() {
        return this.serviceInvoices();
    }
    orders() {
        if (!__classPrivateFieldGet(this, _Bling_orders, "f")) {
            __classPrivateFieldSet(this, _Bling_orders, (0, orders_1.default)(__classPrivateFieldGet(this, _Bling_api, "f"), __classPrivateFieldGet(this, _Bling_raw, "f")), "f");
        }
        return __classPrivateFieldGet(this, _Bling_orders, "f");
    }
    pedidos() {
        return this.orders();
    }
    purchaseOrders() {
        if (!__classPrivateFieldGet(this, _Bling_purchaseOrders, "f")) {
            __classPrivateFieldSet(this, _Bling_purchaseOrders, (0, purchaseOrders_1.default)(__classPrivateFieldGet(this, _Bling_api, "f"), __classPrivateFieldGet(this, _Bling_raw, "f")), "f");
        }
        return __classPrivateFieldGet(this, _Bling_purchaseOrders, "f");
    }
    pedidosDeCompra() {
        return this.purchaseOrders();
    }
    products() {
        if (!__classPrivateFieldGet(this, _Bling_products, "f")) {
            __classPrivateFieldSet(this, _Bling_products, (0, products_1.default)(__classPrivateFieldGet(this, _Bling_api, "f"), __classPrivateFieldGet(this, _Bling_raw, "f")), "f");
        }
        return __classPrivateFieldGet(this, _Bling_products, "f");
    }
    produtos() {
        return this.products();
    }
    commercialProposals() {
        if (!__classPrivateFieldGet(this, _Bling_commercialProposals, "f")) {
            __classPrivateFieldSet(this, _Bling_commercialProposals, (0, commercialProposals_1.default)(__classPrivateFieldGet(this, _Bling_api, "f"), __classPrivateFieldGet(this, _Bling_raw, "f")), "f");
        }
        return __classPrivateFieldGet(this, _Bling_commercialProposals, "f");
    }
    propostasComerciais() {
        return this.commercialProposals();
    }
}
exports.Bling = Bling;
_Bling_api = new WeakMap(), _Bling_raw = new WeakMap(), _Bling_borderos = new WeakMap(), _Bling_customizedFields = new WeakMap(), _Bling_categories = new WeakMap(), _Bling_commercialProposals = new WeakMap(), _Bling_contacts = new WeakMap(), _Bling_deposits = new WeakMap(), _Bling_orders = new WeakMap(), _Bling_products = new WeakMap(), _Bling_purchaseOrders = new WeakMap(), _Bling_invoices = new WeakMap(), _Bling_shopCategories = new WeakMap(), _Bling_billsToPay = new WeakMap(), _Bling_billsToReceive = new WeakMap(), _Bling_contracts = new WeakMap(), _Bling_ctes = new WeakMap(), _Bling_paymentMethods = new WeakMap(), _Bling_productGroups = new WeakMap(), _Bling_nfces = new WeakMap(), _Bling_serviceInvoices = new WeakMap();
